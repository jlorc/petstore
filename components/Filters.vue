<template>
  <div class="filters">
    <div v-click-outside="hideDropdown" class="filters__wrapper">
      <button class="filters__btn" @click="dropdownActive = !dropdownActive">
        {{ currentFilter }}
        <span class="filters__icon" :class="{ active: dropdownActive }"></span>
      </button>
      <div class="filters__dropdown" :class="{ active: dropdownActive }">
        <div
          v-for="filter in filtersOptions"
          class="filters__dropdown-item"
          @click="setFilter(filter)"
        >
          {{ filter }}
          <span
            v-if="filter === currentFilter"
            class="filters__dropdown-icon"
          ></span>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator'
import { namespace } from 'vuex-class'
import vClickOutside from 'v-click-outside'
import {
  PETS_MODULE_ACTIONS,
  PETS_MODULE_GETTERS,
  PETS_MODULE_MUTATIONS,
} from '~/store/constants/PetsModuleConstants'

Vue.use(vClickOutside)

const petsModule = namespace('modules/PetsModule')

@Component
export default class Filters extends Vue {
  @petsModule.Getter(PETS_MODULE_GETTERS.GET_CURRENT_FILTER) currentFilter: string
  @petsModule.Mutation(PETS_MODULE_MUTATIONS.SET_CURRENT_FILTER) setCurrentFilter: (filterStatus: string) => void
  @petsModule.Mutation(PETS_MODULE_MUTATIONS.SET_LOADING) setLoading: (loading: boolean) => void
  @petsModule.Action(PETS_MODULE_ACTIONS.FETCH_PETS_BY_STATUS) fetchPetsByStatus: (status: string) => Promise<string>
  filtersOptions: Array<string> = ['available', 'pending', 'sold']
  dropdownActive: boolean = false

  setFilter(filter: string): void {
    this.setLoading(true)
    this.fetchPetsByStatus(filter).then(() => {
      this.setCurrentFilter(filter)
      this.setLoading(false)
    })
    this.dropdownActive = false
  }

  hideDropdown(): void {
    this.dropdownActive = false
  }
}
</script>

<style lang="scss">
.filters {
  margin: 45px 0 30px;

  &__wrapper {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    position: relative;
  }

  &__icon {
    display: inline-block;
    width: 15px;
    height: 10px;
    background-image: url('@/assets/dropdown.svg');
    background-position: center;
    background-repeat: no-repeat;
    background-size: contain;
    margin-left: 10px;
    transition: 0.25s transform ease;

    &.active {
      transform: rotate(180deg);
    }
  }

  &__dropdown {
    position: absolute;
    top: 46px;
    right: 0;
    max-width: 180px;
    width: 100%;
    visibility: hidden;
    opacity: 0;
    transition: 0.25s opacity ease;
    border: 1px solid $border-color;
    z-index: 3;

    &.active {
      visibility: visible;
      opacity: 1;
    }

    &-item {
      display: flex;
      justify-content: space-between;
      align-items: center;
      padding: 12px 20px;
      font-size: 18px;
      text-transform: capitalize;
      background-color: white;
      width: 100%;
      border-bottom: 1px solid $border-color;
      cursor: pointer;

      &:last-child {
        border-bottom: none;
      }
    }

    &-icon {
      display: inline-block;
      width: 15px;
      height: 10px;
      background-image: url('@/assets/check.svg');
      background-position: center;
      background-repeat: no-repeat;
      background-size: contain;
      margin-left: auto;
    }
  }

  &__btn {
    text-transform: capitalize;
    background-color: white;
    border: 1px solid $border-color;
    font-size: 18px;
    padding: 11px 25px;
    outline: none;
    cursor: pointer;
    border-radius: 3px;
    min-width: 155px;
  }
}
</style>
