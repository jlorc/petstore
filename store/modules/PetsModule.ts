import { Module, VuexModule, Mutation, Action } from 'vuex-module-decorators'
import {
  PETS_MODULE_ACTIONS,
  PETS_MODULE_MUTATIONS,
  PETS_MODULE_GETTERS,
  Pet,
} from '~/store/constants/PetsModuleConstants'

@Module({
  namespaced: true,
  stateFactory: true,
})
export default class PetsModule extends VuexModule {
  pets: Array<Pet> | [] = []
  currentFilter: string = 'available'
  isLoadingActive: boolean = false

  @Action({ commit: PETS_MODULE_MUTATIONS.SET_PETS_BY_STATUS, rawError: true })
  async [PETS_MODULE_ACTIONS.FETCH_PETS_BY_STATUS](
    status: string
  ): Promise<Array<Pet>> {
    const response = await fetch(
      `https://petstore.swagger.io/v2/pet/findByStatus?status=${status}`
    )
    return response.json()
  }

  @Mutation
  [PETS_MODULE_MUTATIONS.SET_PETS_BY_STATUS](petsList: Array<Pet>): void {
    this.pets = petsList
  }

  @Mutation
  [PETS_MODULE_MUTATIONS.SET_CURRENT_FILTER](filterStatus: string): void {
    this.currentFilter = filterStatus
  }

  @Mutation
  [PETS_MODULE_MUTATIONS.SET_LOADING](loading: boolean): void {
    this.isLoadingActive = loading
  }

  get [PETS_MODULE_GETTERS.GET_PETS_BY_STATUS](): Array<Pet> | [] {
    return this.pets
  }

  get [PETS_MODULE_GETTERS.GET_CURRENT_FILTER](): string {
    return this.currentFilter
  }

  get [PETS_MODULE_GETTERS.GET_LOADING](): boolean {
    return this.isLoadingActive
  }
}
