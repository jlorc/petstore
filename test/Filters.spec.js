import { shallowMount, createLocalVue } from '@vue/test-utils'
import Filters from '@/components/Filters.vue'
import * as Vuex from 'vuex'
import PetsModule from '~/store/modules/PetsModule'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Filters', () => {
  let wrapper

  beforeEach(() => {
    const store = new Vuex.Store({
      state: {},
      modules: {
        'modules/PetsModule': {
          ...PetsModule,
        },
      },
    })

    wrapper = shallowMount(Filters, {
      localVue,
      store,
    })
  })

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
