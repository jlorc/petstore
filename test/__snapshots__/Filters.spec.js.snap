// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters renders correctly 1`] = `
<div
  class="filters"
>
  <div
    class="filters__wrapper"
  >
    <button
      class="filters__btn"
    >
      
      available
      
      <span
        class="filters__icon"
      />
    </button>
     
    <div
      class="filters__dropdown"
    >
      <div
        class="filters__dropdown-item"
      >
        
        available
        
        <span
          class="filters__dropdown-icon"
        />
      </div>
      <div
        class="filters__dropdown-item"
      >
        
        pending
        
        <!---->
      </div>
      <div
        class="filters__dropdown-item"
      >
        
        sold
        
        <!---->
      </div>
    </div>
  </div>
</div>
`;
